ir/memory.cpp

```diff
@@ -1009,24 +1009,26 @@ static expr mk_liveness_array() {
   return expr::mkInt(-1, num_nonlocals);
 }
 
-void Memory::mk_nonlocal_val_axioms(bool skip_consts) {
+void Memory::mkNonlocalValAxiom(unsigned i) {
   if (!does_ptr_mem_access)
     return;
 
   expr offset
     = expr::mkFreshVar("#off", expr::mkUInt(0, Pointer::bitsShortOffset()));
 
-  for (unsigned i = has_null_block + skip_consts * num_consts_src,
-       e = numNonlocals(); i != e; ++i) {
-    Byte byte(*this, non_local_block_val[i].val.load(offset));
-    Pointer loadedptr = byte.ptr();
-    expr bid = loadedptr.getShortBid();
-    state->addAxiom(
-      expr::mkForAll({ offset },
-        byte.isPtr().implies(!loadedptr.isLocal(false) &&
-                             !loadedptr.isNocapture(false) &&
-                             bid.ule(numNonlocals() - 1))));
-  }
+  Byte byte(*this, non_local_block_val[i].val.load(offset));
+  Pointer loadedptr = byte.ptr();
+  expr bid = loadedptr.getShortBid();
+
+  // num_nonlocals_src - 1 is reserved as a block for recording fn call's
+  // effects
+  unsigned bid_upperbound =
+      numNonlocals() - 1 - (state->isSource() && has_fncall);
```

Comment:
For the target, `bid != get_fncallmem_bid()` makes the expression bigger; I have no idea if it's strictly better.

Reply:
Isn't it needed for correctness?

Reply:
I think it is okay - a fncallmem is never accessed unless there is a pointer variable in IR that may alias with the fncallmem.
This is already restricted by the precondition in `load` and the pointer arguments, so giving a constraint to fncallmem is duplicated.

Reply:
Okay, I think it is necessary because otherwise the POR precondition can always be false.